File: tugas_hash/hashTABLE.py
class hashNode:
    # Konstruktor node untuk menyimpan key dan value
    def __init__(self, key, value):
        self.key = key
        self.value = value


class hashMap:
    # Konstruktor hash table
    def __init__(self):
        self.capacity = 101
        self.size = 0
        self.arr = [None] * self.capacity
        self.dummy = hashNode(-1, -1)

    # Fungsi hash untuk menentukan index awal
    def hashCode(self, key):
        return abs(key) % self.capacity

    # Menambahkan key-value pair menggunakan linear probing
    def insertNode(self, key, value):
        if key == -1:
            return False, -1, -1

        hashIndex = self.hashCode(key)
        startIndex = hashIndex
        firstDeletedIndex = -1
        counter = 0

        while counter < self.capacity:
            currentNode = self.arr[hashIndex]

            if currentNode is None:
                targetIndex = firstDeletedIndex if firstDeletedIndex != -1 else hashIndex
                self.arr[targetIndex] = hashNode(key, value)
                self.size += 1
                return True, startIndex, targetIndex

            if currentNode.key == -1 and firstDeletedIndex == -1:
                firstDeletedIndex = hashIndex
            elif currentNode.key == key:
                return False, startIndex, hashIndex

            hashIndex = (hashIndex + 1) % self.capacity
            counter += 1

        if firstDeletedIndex != -1:
            self.arr[firstDeletedIndex] = hashNode(key, value)
            self.size += 1
            return True, startIndex, firstDeletedIndex

        return False, startIndex, -1

    # Menghapus data berdasarkan key
    def deleteNode(self, key):
        hashIndex = self.hashCode(key)
        counter = 0

        while self.arr[hashIndex] is not None and counter < self.capacity:
            if self.arr[hashIndex].key == key:
                deletedValue = self.arr[hashIndex].value
                deletedIndex = hashIndex
                self.arr[hashIndex] = self.dummy
                self.size -= 1
                return deletedValue, deletedIndex

            hashIndex = (hashIndex + 1) % self.capacity
            counter += 1

        return -1, -1

    # Mencari value berdasarkan key
    def get(self, key):
        hashIndex = self.hashCode(key)
        startIndex = hashIndex
        counter = 0

        while self.arr[hashIndex] is not None and counter < self.capacity:
            if self.arr[hashIndex].key == key:
                return self.arr[hashIndex].value, startIndex, hashIndex

            hashIndex = (hashIndex + 1) % self.capacity
            counter += 1

        return -1, startIndex, -1

    # Mengembalikan jumlah data dalam hash table
    def sizeofMap(self):
        return self.size

    # Mengecek apakah hash table kosong
    def isEmpty(self):
        return self.size == 0

    # Menampilkan seluruh isi hash table
    def display(self):
        print("\nIsi hash table:")
        for index in range(self.capacity):
            currentNode = self.arr[index]

            if currentNode is not None and currentNode.key != -1:
                print(f"Index {index}: {currentNode.key}")
            else:
                print(f"Index {index}: -")


INITIAL_KEYS = [
    42, 517, 893, 126, 754, 311, 968, 205, 679, 34,
    821, 456, 990, 143, 588, 267, 732, 91, 604, 379,
    850, 12, 699, 234, 941, 118, 563, 807, 326, 471,
    75, 916, 284, 637, 159, 702, 448, 995, 23, 531,
    864, 196, 715, 352, 609, 88, 973, 240, 681, 407,
    129, 556, 814, 367, 925, 51, 493, 770, 218, 645,
    390, 100, 836, 274, 577, 958, 6, 724, 439, 187,
    662, 301, 889, 134, 548, 791, 253, 610, 967, 72,
    415, 830, 199, 703, 346, 581, 912, 29, 468, 755,
    320, 647, 108, 884, 227, 540, 799, 64, 371, 936
]


def loadInitialData(hashTable):
    for key in INITIAL_KEYS:
        hashTable.insertNode(key, key)


def readNumber(message):
    while True:
        try:
            return int(input(message))
        except ValueError:
            print("Input harus berupa angka.")


def showMenu():
    print("\n===== MENU HASH TABLE =====")
    print("1. INPUT DATA")
    print("2. HAPUS DATA")
    print("3. CARI DATA")
    print("4. TAMPILKAN HASH TABLE")
    print("0. KELUAR")


if __name__ == "__main__":
    hashTable = hashMap()
    loadInitialData(hashTable)

    print("Program Hash Table dengan Linear Probing")
    print(f"Data awal sebanyak {hashTable.sizeofMap()} angka random unik sudah dimasukkan.")

    while True:
        showMenu()
        choice = readNumber("Pilih menu: ")

        if choice == 1:
            keyToInsert = readNumber("Masukkan data numerik: ")
            isInserted, hashIndex, storedIndex = hashTable.insertNode(keyToInsert, keyToInsert)

            if isInserted:
                print(
                    f"Data {keyToInsert} berhasil ditambahkan. "
                    f"Hash index: {hashIndex}, stored index: {storedIndex}."
                )
            elif keyToInsert == -1:
                print("Data -1 tidak dapat digunakan karena dipakai sebagai penanda data terhapus.")
            elif storedIndex == -1:
                print(f"Data {keyToInsert} gagal ditambahkan karena hash table sudah penuh.")
            else:
                print(f"Data {keyToInsert} sudah ada pada index {storedIndex}, tidak boleh duplikat.")

        elif choice == 2:
            keyToDelete = readNumber("Masukkan data yang akan dihapus: ")
            deletedValue, deletedIndex = hashTable.deleteNode(keyToDelete)

            if deletedValue != -1:
                print(f"Data {keyToDelete} berhasil dihapus dari index {deletedIndex}.")
            else:
                print(f"Data {keyToDelete} tidak ditemukan.")

        elif choice == 3:
            keyToSearch = readNumber("Masukkan data yang dicari: ")
            value, hashIndex, storedIndex = hashTable.get(keyToSearch)

            if value != -1:
                print(
                    f"Data {keyToSearch} ditemukan. "
                    f"Hash index: {hashIndex}, stored index: {storedIndex}."
                )
            else:
                print(f"Data {keyToSearch} tidak ditemukan. Hash index awal: {hashIndex}.")

        elif choice == 4:
            hashTable.display()

        elif choice == 0:
            print("Program selesai.")
            break

        else:
            print("Pilihan tidak tersedia.")

File: tugas_hash/src/hash_table.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Scanner;

public class hash_table {
    private static final int DEFAULT_BUCKET_COUNT = 101;

    private static final int[] INITIAL_KEYS = {
        42, 517, 893, 126, 754, 311, 968, 205, 679, 34,
        821, 456, 990, 143, 588, 267, 732, 91, 604, 379,
        850, 12, 699, 234, 941, 118, 563, 807, 326, 471,
        75, 916, 284, 637, 159, 702, 448, 995, 23, 531,
        864, 196, 715, 352, 609, 88, 973, 240, 681, 407,
        129, 556, 814, 367, 925, 51, 493, 770, 218, 645,
        390, 100, 836, 274, 577, 958, 6, 724, 439, 187,
        662, 301, 889, 134, 548, 791, 253, 610, 967, 72,
        415, 830, 199, 703, 346, 581, 912, 29, 468, 755,
        320, 647, 108, 884, 227, 540, 799, 64, 371, 936
    };

    // Jumlah bucket pada hash table
    private int bucketCount;

    // Setiap bucket berisi linked list untuk menangani collision
    private List<List<Integer>> table;

    private int dataCount;

    public hash_table(int buckets) {
        bucketCount = buckets;
        dataCount = 0;
        table = new ArrayList<>();

        for (int index = 0; index < bucketCount; index++) {
            table.add(new LinkedList<>());
        }
    }

    public boolean insert(int key) {
        if (search(key)) {
            return false;
        }

        int index = getHashIndex(key);
        table.get(index).add(key);
        dataCount++;
        return true;
    }

    public boolean remove(int key) {
        int index = getHashIndex(key);
        boolean isRemoved = table.get(index).remove(Integer.valueOf(key));

        if (isRemoved) {
            dataCount--;
        }

        return isRemoved;
    }

    public boolean search(int key) {
        int index = getHashIndex(key);
        return table.get(index).contains(key);
    }

    public int getIndexIfFound(int key) {
        int index = getHashIndex(key);

        if (table.get(index).contains(key)) {
            return index;
        }

        return -1;
    }

    public int getChainPositionIfFound(int key) {
        int index = getHashIndex(key);
        List<Integer> chain = table.get(index);

        for (int position = 0; position < chain.size(); position++) {
            if (chain.get(position).equals(key)) {
                return position + 1;
            }
        }

        return -1;
    }

    public void display() {
        System.out.println("\nIsi hash table:");
        for (int index = 0; index < bucketCount; index++) {
            System.out.print(index);

            for (int key : table.get(index)) {
                System.out.print(" --> " + key);
            }

            System.out.println();
        }
    }

    public int getDataCount() {
        return dataCount;
    }

    private int getHashIndex(int key) {
        return Math.abs(key) % bucketCount;
    }

    private static void loadInitialData(hash_table hashTable) {
        for (int key : INITIAL_KEYS) {
            hashTable.insert(key);
        }
    }

    private static int readNumber(Scanner input, String message) {
        while (true) {
            System.out.print(message);
            if (input.hasNextInt()) {
                return input.nextInt();
            }

            System.out.println("Input harus berupa angka.");
            input.next();
        }
    }

    private static void showMenu() {
        System.out.println("\n===== MENU HASH TABLE =====");
        System.out.println("1. INPUT DATA");
        System.out.println("2. HAPUS DATA");
        System.out.println("3. CARI DATA");
        System.out.println("4. TAMPILKAN HASH TABLE");
        System.out.println("0. KELUAR");
    }

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        hash_table hashTable = new hash_table(DEFAULT_BUCKET_COUNT);
        loadInitialData(hashTable);

        System.out.println("Program Hash Table dengan Separate Chaining");
        System.out.println("Data awal sebanyak " + hashTable.getDataCount() + " angka random unik sudah dimasukkan.");

        int choice;
        do {
            showMenu();
            choice = readNumber(input, "Pilih menu: ");

            switch (choice) {
                case 1:
                    int keyToInsert = readNumber(input, "Masukkan data numerik: ");
                    if (hashTable.insert(keyToInsert)) {
                        System.out.println("Data " + keyToInsert + " berhasil ditambahkan.");
                    } else {
                        System.out.println("Data " + keyToInsert + " sudah ada, tidak boleh duplikat.");
                    }
                    break;
                case 2:
                    int keyToRemove = readNumber(input, "Masukkan data yang akan dihapus: ");
                    if (hashTable.remove(keyToRemove)) {
                        System.out.println("Data " + keyToRemove + " berhasil dihapus.");
                    } else {
                        System.out.println("Data " + keyToRemove + " tidak ditemukan.");
                    }
                    break;
                case 3:
                    int keyToSearch = readNumber(input, "Masukkan data yang dicari: ");
                    int foundIndex = hashTable.getIndexIfFound(keyToSearch);
                    if (foundIndex != -1) {
                        int chainPosition = hashTable.getChainPositionIfFound(keyToSearch);
                        System.out.println("Data " + keyToSearch + " ditemukan pada index "
                                + foundIndex + ", posisi linked list ke-" + chainPosition + ".");
                    } else {
                        System.out.println("Data " + keyToSearch + " tidak ditemukan.");
                    }
                    break;
                case 4:
                    hashTable.display();
                    break;
                case 0:
                    System.out.println("Program selesai.");
                    break;
                default:
                    System.out.println("Pilihan tidak tersedia.");
                    break;
            }
        } while (choice != 0);

        input.close();
    }
}
